### ISSUE TITLE
Invalid ScissorRect parameters in examples

### ISSUE THREAD

**Issue by dzil123** [Yatekii/imgui-wgpu-rs#53]
Affects 97687c2.
I get this error instantly on startup when I clone this repo and run any example, except the two simple_api examples.
There is also an out-of-bounds validation error from Vulkan.

Related: issue #48, except this issue is distinct because it affects the examples.

```
wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 1, Vulkan)>`
    In a set_scissor_rect command
    Invalid ScissorRect parameters
```

<details><summary>Full traceback</summary><p>

```
$ RUST_BACKTRACE=1 cargo run --example hello_world    
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `target/debug/examples/hello_world`
[0.069024 ERROR]()(no module): 
VALIDATION [VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 (2094043421)] : Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x559b78036f30, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (800,600), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (1387,780), minImageExtent = (1387,780), maxImageExtent = (1387,780). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
object info: (type: DEVICE, hndl: 94126221782832)

wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 1, Vulkan)>`
    In a set_scissor_rect command
    Invalid ScissorRect parameters


thread 'main' panicked at 'Handling wgpu errors as fatal by default', ~/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.7.1/src/backend/direct.rs:1896:5
stack backtrace:
   0: std::panicking::begin_panic
             at ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:519:12
   1: wgpu::backend::direct::default_error_handler
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.7.1/src/backend/direct.rs:1896:5
   2: core::ops::function::Fn::call
             at ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:70:5
   3: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1535:9
   4: wgpu::backend::direct::ErrorSinkRaw::handle_error
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.7.1/src/backend/direct.rs:1883:9
   5: wgpu::backend::direct::Context::handle_error
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.7.1/src/backend/direct.rs:109:9
   6: <wgpu::backend::direct::Context as wgpu::Context>::command_encoder_end_render_pass
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.7.1/src/backend/direct.rs:1739:13
   7: <wgpu::RenderPass as core::ops::drop::Drop>::drop
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.7.1/src/lib.rs:2486:13
   8: core::ptr::drop_in_place<wgpu::RenderPass>
             at ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:179:1
   9: core::mem::drop
             at ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/mod.rs:895:24
  10: hello_world::main::{{closure}}
             at ./examples/hello_world.rs:219:17
  11: winit::platform_impl::platform::sticky_exit_callback
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/mod.rs:736:5
  12: winit::platform_impl::platform::x11::EventLoop<T>::run_return
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/x11/mod.rs:330:17
  13: winit::platform_impl::platform::x11::EventLoop<T>::run
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/x11/mod.rs:398:9
  14: winit::platform_impl::platform::EventLoop<T>::run
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/mod.rs:652:56
  15: winit::event_loop::EventLoop<T>::run
             at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/event_loop.rs:154:9
  16: hello_world::main
             at ./examples/hello_world.rs:106:5
  17: core::ops::function::FnOnce::call_once
             at ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[0.125825 ERROR]()(gpu_descriptor::allocator): `DescriptorAllocator` is dropped while some descriptor sets were not deallocated
[0.125940 ERROR]()(gpu_alloc::block): Memory block wasn't deallocated
[0.125956 ERROR]()(gpu_alloc::block): Memory block wasn't deallocated
[0.125970 ERROR]()(gpu_alloc::block): Memory block wasn't deallocated
[0.125992 ERROR]()(gpu_alloc::block): Memory block wasn't deallocated
[0.126007 ERROR]()(gpu_alloc::block): Memory block wasn't deallocated
[0.126024 ERROR]()(gpu_alloc::block): Memory block wasn't deallocated
[0.126042 ERROR]()(gpu_alloc::block): Memory block wasn't deallocated
[0.126062 ERROR]()(gpu_alloc::block): Memory block wasn't deallocated
Segmentation fault (core dumped)
```

</p></details>

**Comment by JMS55** [comment]
I just ran into the same issue in my actual game, when resizing the OS window slowly and causing an imgui window to be big enough that a scrollbar is no longer needed.

**Comment by 4bb4** [comment]
I have the same issue over in https://github.com/4bb4/implot-rs/issues/13, just posting here so I get notifications if anything comes to light :) I've been wondering if it's an implot-specific issue, but it seems to come up in other places too.

**Comment by mizkichan** [comment]
Though this is just a workaround, passing `HiDpiMode::Locked(1.0)` to `WinitPlatform::attach_window` prevents the error for me.

**Comment by JMS55** [comment]
I tried that, didn't fix it for me.

**Comment by JMS55** [comment]
I did some debugging, and it turns out that wgpu is incorrectly saying 0 width/height scissors are invalid. This is the cause of the majority of the crashes.

Additionally, there seems to be some sort of desync between what imgui thinks the attachment texture size is, and what it actually is. This results in scissors with x+width/y+height greater than the attachment size, which is definitely invalid. We're going to need to track down why the desync happens. I'm able to trigger this by rapidly resizing the window back and forth a small amount.

**Comment by dzil123** [comment]
I tried the examples again on 787a720 and it still crashes instantly on startup. It happen instantly for the examples, but in other projects it only happens after some resizing or moving an imgui window out of view. 

I don't know if this is relevant, but theres a new start to the output, before the wgpu ScissorRect validation error:

```
[0.108018 ERROR]()(no module): 
VALIDATION [VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 (0x7cd0911d)] : Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x56467eeb7220, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (800,600), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (1387,780), minImageExtent = (1387,780), maxImageExtent = (1387,780). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
object info: (type: DEVICE, hndl: 0x56467eeb7220)

[0.110609 ERROR]()(no module): Unexpected varying type: Array { base: [1], size: Constant([5]), stride: 4 }
```

**Comment by michael-swan** [comment]
So I tried making a patch to fix this ScissorRect problem and while it averts a crash, I'm noticing that when this crash averting code is frequently hit, an insane amount of memory is leaked. e.g. it averted this crash 68 times in my current run over the course of maybe a minute and now the process is allocating 5.7 GB's of memory. It was only consuming 200 MB's prior.

**Comment by 4bb4** [comment]
> I did some debugging, and it turns out that wgpu is incorrectly saying 0 width/height scissors are invalid. This is the cause of the majority of the crashes.

I checked and this is also the case for my issues over in `implot-rs`. Interestingly enough, if I locally patch the check for 0 out of `wgpu-core`, the crashes go away ('ve not observed the memory leaks @michael-swan saw, but I guess that's because I removed the check [here](https://github.com/gfx-rs/wgpu/blob/master/wgpu-core/src/command/render.rs#L1285) so the code actually still goes through all the motions).

While getting rid of the crashes (which for me happen when I scroll in a window with many collapsible headers) is great, I then see some other strange behavior that might have to do with the desync mentioned by @JMS55: If I scroll the scroll bar to _just the right location_, I get a visual glitch where the imgui window contents are drawn with different vertical offsets in alternating frames. If I create a screenshot or even a video capture, the flashing does not seem to get recorded, but I filmed it with a high-frame-rate camera and could confirm that the frames indeed alternate:
![image](https://user-images.githubusercontent.com/67376761/124365990-65e40680-dc4c-11eb-86cc-bfb0b2e5a4c3.png)


As always, I'm absolutely not sure if this funky behavior is an error on my end. I didn't have it with older versions of WGPU, imgui-wgpu-rs and imgui, but that could be a red herring. I'm just providing the info here in case it helps.
